scrapy/settings/__init__.py

```diff
@@ -34,9 +40,20 @@ def __init__(self, value, priority):
 
     def set(self, value, priority):
         """Sets value if priority is higher or equal than current priority."""
-        if priority >= self.priority:
-            self.value = value
-            self.priority = priority
+        if isinstance(self.value, BaseSettings) and isinstance(value, Mapping):
+            # Ignore self.priority if self.value has per-key priorities
+            # XXX: Second condition allows completely replacing a BaseSettings
+            #      instance with a value of incompatible type (i.e. something
+            #      that does not map keys to values, e.g.a list). I included
+            #      this b/c the test_cmdline/__init__.py test uses a deprecated
+            #      configuration API (EXTENSIONS is a list) and fails otherwise.
+            #      Should it stay in?
+            self.value.update(value, priority)
+            self.priority = max((priority, self.priority))
```

Comment:
Yep, that was the idea. I've updated the code so it works as expected, thereby introducing a `maxpriority()` method to the `BaseSettings` class since there are two places in the code where I need to find the maximum priority. It's probably not very useful for anything else though, does it clutter up the API too much?